### ISSUE TITLE
Customization and shutdown of updater's executor

### ISSUE THREAD

**Issue by netvl** [kotlin-telegram-bot/kotlin-telegram-bot#127]
**Is your feature request related to a problem? Please describe.**

At the moment, the polling happens in a fixed-configuration thread pool:

```kotlin
class Updater {
    private val executor: Executor = Executors.newCachedThreadPool()
    ...
}
```

This might not be a big issue for many applications, but in some cases it is very useful or even necessary to substitute or otherwise customize the underlying executor, for example, when the library is used as a part of bigger project with an already existing executor.

What's even more important, the library currently does not expose any shutdown methods for this thread pool, and because it uses default configuration for the cached thread pool, all threads that this pool contains are non-daemon. This means that there is no way to shut down the application using the library cleanly "from inside" of it, because once the update process starts, the thread pool will start its threads, and since they are non-daemon, shutting down all other threads in a program will not result in the JVM shutdown (although it seems to me that the default configuration of the cached thread pool will eventually shut down all idle threads, allowing the application to stop, but this delay unfortunately cannot be controlled by the user). Again, this behavior is highly inconvenient in applications which the bot is only a part of.

**Describe the solution you'd like**

Ideally, the library API should allow explicitly setting the updater's executor service, maybe something like

```kotlin
val customExecutor = ...

bot {
    updater.executor = customExecutor
    ...
}
```

At the very least, there should be some kind of `shutdown()` method, either on `Bot` or on `Updater`, which would invoke the underlying executor's `shutdown()` routine, in order to allow shutting down the default executor service:

```
// Either on `Updater` or on `Bot`

fun shutdown() {
    if (executor is ExecutorService) {
        executor.shutdownNow()
    }
}

fun awaitTermination() {
    if (executor is ExecutorService) {
        executor.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS)
    }
}

fun shutdownNow() {
    shutdown()
    awaitTermination()
}
```

**Describe alternatives you've considered**

There does not seem to be any alternatives, except very hacky ones like using reflection to extract the reference to the `Executor`.

**Implementation ideas**

Described above.

**Additional context**

N/A

**Comment by vjgarciag96** [comment]
Hey @netvl! 

I totally agree with you, the current implementation using a non-customisable thread pool and not exposing an API to shutdown the threads is not the ideal thing to have.

Also, even when using web hooks the library creates an additional thread, so there is no way to avoid using the library executor nowadays.

I thought about all this but didn't have enough time to really materialise some of the ideas.

Currently, if we run the bot with polling it uses two threads, one for the polling against Telegram servers and one for a loop in which we take updates from a blocking queue and dispatch them (and we don't block the "main" thread or the thread that initiates the bot).

If we run the bot with web hooks it still uses the thread to take updates from the blocking queue and dispatch them (and again, we don't block the "main" thread or the thread used to initiate the bot).

I was thinking on using by default the thread that initiates the bot to do the "take updates from the queue and dispatch them" job and providing an option to choose the thread you want that to run on in case you don't want to block the "initiator thread".

I was also thinking on allowing the customisation of the thread used for the polling against Telegram servers.

It would be nice to also provide the shutdown API you sugessted.

Any thoughts, feedback or additional ideas on this?

**Comment by netvl** [comment]
> I was thinking on using by default the thread that initiates the bot to do the "take updates from the queue and dispatch them" job and providing an option to choose the thread you want that to run on in case you don't want to block the "initiator thread".
>
> I was also thinking on allowing the customisation of the thread used for the polling against Telegram servers.

Yes, that's pretty much what I would have expected of such a library, and this approach would give the most control to the user of the library (they will be able to choose both the thread where the primary logic of the bot runs, and they will be able to choose where the polling logic is executed).
